src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs

```diff
@@ -162,11 +188,39 @@ public override void HandleAssignment (MultiValue source, MultiValue target, IOp
 			);
 		}
 
-		public override MultiValue HandleArrayElementAccess (IOperation arrayReferene)
+		public override MultiValue HandleArrayElementRead (MultiValue arrayValue, MultiValue indexValue, IOperation operation)
 		{
+			if (arrayValue.AsSingleValue () is not ArrayValue arr)
+				return UnknownValue.Instance;
+
+			if (indexValue.AsConstInt () is not int index)
+				return UnknownValue.Instance;
+
+			if (arr.TryGetValueByIndex (index, out var elementValue))
+				return elementValue;
+
 			return UnknownValue.Instance;
 		}
 
+		public override void HandleArrayElementWrite (MultiValue arrayValue, MultiValue indexValue, MultiValue valueToWrite, IOperation operation)
+		{
+			int? index = indexValue.AsConstInt ();
+			foreach (var arraySingleValue in arrayValue) {
+				if (arraySingleValue is ArrayValue arr) {
+					if (index == null) {
+						// Reset the array to all unknowns - since we don't know which index is being assigned
+						arr.IndexValues.Clear ();
+					} else {
+						if (arr.IndexValues.TryGetValue (index.Value, out var existingValue)) {
+							arr.IndexValues[index.Value] = _multiValueLattice.Meet (existingValue, valueToWrite);
```

Comment:
Hmm - in the analyzer you're probably right - it should overwrite. Linker is different since it doesn't have the solver yet, so it tracks basic blocks - but it does overwrite withing the same basic block.